File: custom_components/rts_link/api/ha_api.py
from homeassistant.components.http.data_validator import RequestDataValidator
from homeassistant.helpers.http import HomeAssistantView
from homeassistant.helpers import config_validation as cv
import voluptuous as vol

from custom_components.rts_link.cover import CoverType
from custom_components.rts_link.rts_link_api import RTSLinkApi
from custom_components.rts_link.const import DOMAIN, RTS_API


class RTSLinkNewProgView(HomeAssistantView):
    """Endpoint to adding new cover."""

    url = "/api/rts_link/cover/new"
    name = "api:rts_link:cover:new"

    @RequestDataValidator(
        vol.Schema(
            {
                vol.Required('name'): cv.string,
                vol.Required('type'): vol.In(CoverType),
            }
        )
    )
    async def post(self, request, data):
        hass = request.app["hass"]
        rts_api: RTSLinkApi = hass.data[DOMAIN][RTS_API]
        result = await rts_api.add_cover(data['name'], data['type'])
        return self.json({"success": result})


class RTSLinkAddShutterView(HomeAssistantView):
    """Endpoint to adding new shutter."""

    url = "/api/rts_link/cover/add"
    name = "api:rts_link:cover:add"

    @RequestDataValidator(
        vol.Schema(
            {
                vol.Required('id'): cv.string,
            }
        )
    )
    async def post(self, request, data):
        hass = request.app["hass"]
        rts_api: RTSLinkApi = hass.data[DOMAIN][RTS_API]
        result = await rts_api.add_shutter_to_existing_cover(int(data['id']))
        return self.json({"success": result})


class RTSLinkRemoveCoverView(HomeAssistantView):
    """Endpoint to remove a cover."""

    url = "/api/rts_link/cover/remove"
    name = "api:rts_link:cover:remove"

    @RequestDataValidator(
        vol.Schema(
            {
                vol.Required('id'): cv.string,
            }
        )
    )
    async def post(self, request, data):
        hass = request.app["hass"]
        rts_api: RTSLinkApi = hass.data[DOMAIN][RTS_API]
        result = await rts_api.remove_cover(int(data['id']))
        return self.json({"success": result})


class RTSLinkRenameCoverView(HomeAssistantView):
    """Endpoint to remove a cover."""

    url = "/api/rts_link/cover/rename"
    name = "api:rts_link:cover:rename"

    @RequestDataValidator(
        vol.Schema(
            {
                vol.Required('id'): cv.string,
                vol.Required('name'): cv.string,
            }
        )
    )
    async def post(self, request, data):
        hass = request.app["hass"]
        rts_api: RTSLinkApi = hass.data[DOMAIN][RTS_API]
        result = await rts_api.rename_cover(int(data['id']), data['name'])
        return self.json({"success": result})


class RTSLinkChangeTypeCoverView(HomeAssistantView):
    """Endpoint to change the type of cover."""

    url = "/api/rts_link/cover/type"
    name = "api:rts_link:cover:type"

    @RequestDataValidator(
        vol.Schema(
            {
                vol.Required('id'): cv.string,
                vol.Required('type'): vol.In(CoverType),
            }
        )
    )
    async def post(self, request, data):
        hass = request.app["hass"]
        rts_api: RTSLinkApi = hass.data[DOMAIN][RTS_API]
        result = await rts_api.change_type_cover(int(data['id']), data['type'])
        return self.json({"success": result})


async def async_register_api(hass):
    hass.http.register_view(RTSLinkNewProgView)
    hass.http.register_view(RTSLinkAddShutterView)
    hass.http.register_view(RTSLinkRemoveCoverView)
    hass.http.register_view(RTSLinkRenameCoverView)
    hass.http.register_view(RTSLinkChangeTypeCoverView)

File: custom_components/rts_link/frontend/src/panel/api/ha-api.ts
import { type HomeAssistant } from 'custom-card-helpers'
import { CoverDeviceEnum } from './enum/cover-device-enum';

interface ApiResponse {
  success: boolean
}

export const rtsLinkNewCover = async (hass: HomeAssistant, name: string, type: CoverDeviceEnum): Promise<ApiResponse> => {
  return await hass.callApi<ApiResponse>('POST', 'rts_link/cover/new', { name: name, type: type })
}

export const rtsLinkAddShutter = async (hass: HomeAssistant, rtsId: number): Promise<ApiResponse> => {
  return await hass.callApi<ApiResponse>('POST', 'rts_link/cover/add', { id: rtsId })
}

export const rtsLinkRemoveCover = async (hass: HomeAssistant, rtsId: number): Promise<ApiResponse> => {
  return await hass.callApi<ApiResponse>('POST', 'rts_link/cover/remove', { id: rtsId })
}

export const rtsLinkRenameCover = async (hass: HomeAssistant, rtsId: number, name: string): Promise<ApiResponse> => {
  return await hass.callApi<ApiResponse>('POST', 'rts_link/cover/rename', { id: rtsId, name: name })
}

export const rtsLinkChangeTypeCover = async (hass: HomeAssistant, rtsId: number, type: CoverDeviceEnum): Promise<ApiResponse> => {
  return await hass.callApi<ApiResponse>('POST', 'rts_link/cover/type', { id: rtsId, type: type })
}
